nixos/modules/services/misc/sourcehut/default.nix

```diff
@@ -0,0 +1,173 @@
+        Enable sourcehut - git hosting, continuous integration, mailing list, ticket tracking,
+        task dispatching, wiki and account management services.
+      '';
+    };
+
+    services = mkOption {
+      type = types.nonEmptyListOf (types.enum [ "builds" "dispatch" "git" "hub" "hg" "lists" "man" "meta" "paste" "todo" ]);
+      default = [ "builds" "dispatch" "git" "hub" "hg" "lists" "man" "meta" "paste" "todo" ];
+      description = ''
+        Services to enable on the sourcehut network.
+      '';
+    };
+
+    address = mkOption {
+      type = types.str;
+      default = "127.0.0.1";
+      description = ''
+      '';
+    };
+
+    python = mkOption {
+      internal = true;
+      type = types.package;
+      default = python;
+      description = ''
+        The python package to use. It should contain references to the *srht modules and also
+        gunicorn.
+      '';
+    };
+
+    statePath = mkOption {
+      type = types.path;
+      default = "/var/sourcehut";
+      description = ''
+        Root state path for the sourcehut network.
+      '';
+    };
+
+    settings = mkOption {
+      type = with types; attrsOf (attrsOf (nullOr (either bool (either int (either float (either str path))))));
+      default = {};
+      description = ''
+        The configuration for the sourcehut network.
+      '';
+    };
+  };
+
+  config = mkIf cfg.enable {
+    assertions =
+      [
+        { assertion = with cfgIni."sr.ht"; secret-key != null && stringLength secret-key == 32;
+          message = "sr.ht's secret key must be defined and of a 32 byte length."; }
+
+        # Is it always 44 characters...? At least from the times I've generated one...
+        { assertion = with cfgIni.webhooks; private-key != null && stringLength private-key == 44;
+          message = "The webhook's private key must be defined."; }
+
+        { assertion = hasAttrByPath [ "meta.sr.ht" "origin" ] cfgIni && cfgIni."meta.sr.ht".origin != null;
+          message = "meta.sr.ht's origin must be defined."; }
+      ];
+
+    virtualisation.docker.enable = true;
+    environment.etc."sr.ht/config.ini".text = let
+      mkKeyValue = key: v: let
+        isPath = v: builtins.typeOf v == "path";
+
+        value =
+          if null == v        then ""
+          else if true == v   then "yes"
+          else if false == v  then "no"
+          else if isInt v     then toString v
+          else if isString v  then toString v
+          else if isPath v    then toString v
+          else abort "sourcehut.mkKeyValue: unexpected type (v = ${v})";
+      in "${toString key}=${value}";
+    in generators.toINI { inherit mkKeyValue; } cfg.settings;
+
+    environment.systemPackages = [ python ];
+
+    # PostgreSQL server
+    services.postgresql.enable = true;
+    # Mail server
+    services.postfix.enable = mkOverride 999 true;
+    # Cron daemon
+    services.cron.enable = mkOverride 999 true;
+    # Redis server
+    services.redis = {
```

Comment:
Does this overwrite my redis settings?

Reply:
Don't know. What's the right merge method here?

Reply:
It should be a conditional right?
somethink like `if config.services.redis.enable then config.service.redis else { ... }`

Reply:
By the same token though, wouldn't that also happen in various places for the postgres config.